### ISSUE TITLE
how to extract feature from the customized layer once I fine tune my model? I am in trouble. Could you please help me?

### ISSUE THREAD

**Issue by csitaula** [rstudio/keras#551]
I have written a code for pre-trained network and fine tuned with my dataset. I want to extract the features from the layer I added. I am trying to extract the features from layer_dense(units = 1024, activation = 'relu') in my fine tuned model. Could anyone help me to extract the features from that custom layer after fine tuning? Thanks

create the base pre-trained model
base_model <- application_inception_v3(weights = 'imagenet', include_top = FALSE)

add our custom layers
predictions <- base_model$output %>%
layer_global_average_pooling_2d() %>%
layer_dense(units = 1024, activation = 'relu') %>%
layer_dense(units = 7, activation = 'softmax')

this is the model we will train
model <- keras_model(inputs = base_model$input, outputs = predictions)

first: train only the top layers (which were randomly initialized)
i.e. freeze all convolutional InceptionV3 layers
freeze_weights(base_model)

compile the model (should be done after setting layers to non-trainable)
model %>% compile(optimizer = 'rmsprop', loss = 'categorical_crossentropy')
train_datagen = image_data_generator(
rescale = 1/255,
rotation_range = 40,
width_shift_range = 0.2,
height_shift_range = 0.2,
shear_range = 0.2,
zoom_range = 0.2,
horizontal_flip = TRUE,
fill_mode = "nearest"
)
test_datagen <- image_data_generator(rescale = 1/255)
train_generator <- flow_images_from_directory(
'd:/nyu/train',
train_datagen,
target_size = c(224, 224),
batch_size = 20,
class_mode = "sparse"
)
test_generator <- flow_images_from_directory(
'd:/nyu/test',
test_datagen,
target_size = c(224, 224),
batch_size = 20,
class_mode = "sparse"
)

model %>% compile(
loss = "binary_crossentropy",
optimizer = optimizer_rmsprop(lr = 2e-5),
metrics = c("accuracy")
)
history <- model %>% fit_generator(
train_generator,
steps_per_epoch = 10,
epochs = 3,
validation_data = test_generator,
validation_steps = 5
)
And the link of dataset is:
https://drive.google.com/open?id=10RKbXaUDYvMNtwiIgdVN1h6Kt8oTa1j0
Hope to hear from you soon! Thanks!

**Comment by skeydan** [comment]
Hi,

you can create a second model that has as output a layer from the trained model, like in this simplified example:

```
input <- layer_input(shape = 4)
dense <- layer_dense(input, units = 32) %>% 
  layer_dense(units = 8) %>% 
  layer_dense(1)

model <- keras_model(inputs = input, outputs = dense)

model %>% compile(optimizer = "adam", loss = "mse")

model %>% fit(
  x = matrix(1:8, nrow = 2),
  y = matrix(1:2, nrow = 2),
  epochs = 1,
  batch_size = 2
)

model$layers

model2 <- keras_model(inputs = input, outputs = model$layers[[3]]$output)
model2

model2 %>% predict(matrix(1:8, nrow = 2))
```

**Comment by csitaula** [comment]
Thanks for this wonderful code! I have created a code like this! and tested the image. Does this way work? 

library(keras)
# create the base pre-trained model
base_model <- application_vgg16(weights = 'imagenet', include_top = FALSE,input_shape = c(224,224,3))
#freeze only layers before 4 layer
layers <- base_model$layers
print(layers)

for(i in layers[1:19])
{
  i$trainable <- FALSE
}
#display if it works or not

for(l in layers)
{
  print(paste(l,l$trainable))
}
model <- keras_model_sequential() %>% 
  base_model %>%
  layer_flatten() %>%
 # layer_dropout(rate = 0.25) %>% 
  layer_dense(units = 1024,activation = 'relu') %>%
  layer_dropout(rate = 0.50) %>% 
  layer_dense(units = 2,activation = 'softmax')
# add our custom layers
# predictions <- base_model$output %>% 
#   layer_global_max_pooling_2d() %>% 
#   layer_dense(units = 1024, activation = 'relu') %>% 
#   layer_dropout(rate = 0.5) %>% 
#   layer_dense(units = 2, activation = 'softmax')
# 
# # this is the model we will train
# model <- keras_model(inputs = base_model$input, outputs = predictions)

# first: train only the top layers (which were randomly initialized)
# i.e. freeze all convolutional InceptionV3 layers
#freeze_weights(base_model)


# compile the model (should be done *after* setting layers to non-trainable)
model %>% compile(optimizer = 'rmsprop', loss = 'categorical_crossentropy')
batch_size<-5
train_samples<-498
validation_samples<-177

train_datagen = image_data_generator(
  rescale = 1./255,
  rotation_range = 20,
  width_shift_range = 0.2,
  height_shift_range = 0.2,
  shear_range = 0.2,
  zoom_range = 0.2,
  horizontal_flip = TRUE,
  fill_mode = "nearest"
)
test_datagen <- image_data_generator(rescale = 1./255)
train_generator <- flow_images_from_directory(
  'C://Experiment/Fine-tune/Train',
  train_datagen,
  target_size = c(224, 224),
  batch_size = batch_size,
  class_mode = "binary"
)
test_generator <- flow_images_from_directory(
  'C:/Experiment/Fine-tune/Test',
  test_datagen,
  target_size = c(224, 224),
  batch_size = batch_size,
  class_mode = "binary"
)

model %>% compile(
  loss = "categorical_crossentropy",
  optimizer = optimizer_rmsprop(lr = 1e-5),
  metrics = c("accuracy")
  
)
history <- 
  model %>% fit_generator(
    train_generator,
    steps_per_epoch = as.integer(train_samples/batch_size),
    epochs = 30,
    #batch_size=5,
    validation_data = test_generator,
    validation_steps = as.integer(validation_samples/batch_size)
  )
 save_model_hdf5(model,'C://Experiment/picalert_fine_tuned_model.h5')
plot(history)

#fine_tuned_model <- load_model_weights_hdf5_hdf5('C://Experiment/picalert_fine_tuned_model.h5')
#visually interpret error and loss
# Plot the model loss of the training data
plot(history$metrics$loss, main="Model Loss", xlab = "epoch", ylab="loss", col="blue", type="l")

# Plot the model loss of the test data
lines(history$metrics$val_loss, col="green")

# Add legend
legend("topright", c("train","test"), col=c("blue", "green"), lty=c(1,1))
#display for accuracy
# Plot the accuracy of the training data 
plot(history$metrics$acc, main="Model Accuracy", xlab = "epoch", ylab="accuracy", col="blue", type="l")

# Plot the accuracy of the validation data
lines(history$metrics$val_acc, col="green")

# Add Legend
legend("bottomright", c("train","test"), col=c("blue", "green"), lty=c(1,1))

#test the image

base_model <- application_vgg16(weights = 'imagenet', include_top = FALSE,input_shape = c(224,224,3))
model <- keras_model_sequential() %>% 
  base_model %>%
  layer_flatten() %>%
 # layer_dropout(rate = 0.25) %>% 
  layer_dense(units = 1024,activation = 'relu') %>%
  layer_dropout(rate = 0.50) %>% 
  layer_dense(units = 2,activation = 'softmax')
  model%>%load_model_weights_hdf5('C://Experiment/picalert_fine_tuned_model.h5')
img_path <- 'C:/Experiment/PicAlertData/private/4314604989.jpg'
#img_path <- "elephant.jpg"
img <- image_load(img_path, target_size = c(224,224))
x <- image_to_array(img)
x<-x/255 #this is for scaling of the image which we had done while traininig
x <- array_reshape(x, c(1, dim(x)))
x <- imagenet_preprocess_input(x)
output <- model %>% predict(x)
output

**Comment by csitaula** [comment]
Thanks a lot!
